Sprint 15

roudYStorm opened the pull request:
посмотрите пожалуйста файл прочитай пожалуйста. Очень буду благодарна если скажете как это все исправить..Простите меня еще раз..

roudYStorm committed:
1

roudYStorm committed:
спринт 15

roudYStorm committed:
цц

roudYStorm committed:
yfxf

IgorChumikov commented on CoreData/Converters/ScheduleConvertor.swift:
```diff
@@ -0,0 +1,48 @@
+        return bits
+    }
+    
+    private func convertFromUInt16(from value: Int16) -> [Int] {
```
Лучше переименовать convertFromUInt16(from:) → convertFromUInt16(_:)
Повторение from from в названии и параметре не нужно.

Swift-комьюнити рекомендует избегать from: у внутренних приватных методов без необходимости.
```
private func convertFromUInt16(_ value: Int16) -> [Int]

IgorChumikov commented on CoreData/Converters/ScheduleConvertor.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+    
+    private func convertFromUInt16(from value: Int16) -> [Int] {
+        let bitsInByte = bits(fromByte: value)
```
можно использовать map, что будет короче и читаемее:
```
return bitsInByte.map { $0.description }

IgorChumikov commented on CoreData/Converters/ScheduleConvertor.swift:
```diff
@@ -0,0 +1,48 @@
+        return weekdays
+    }
+    
+    func convertScheduleToUInt16(from weekdays: Set<WeekDays>) -> Int16 {
```
Можно более современно через reduce(into:):
```
weekdays.reduce(into: Int16(0)) { result, day in
    result |= 1 << day.rawValue
}
```
(Но это по желанию — вариант с циклом тоже норм, особенно для читаемости.)

IgorChumikov commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,175 @@
+// MARK: - NSFetchedResultsControllerDelegate
+extension TrackerCategoryStore: NSFetchedResultsControllerDelegate {
+    func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
+        insertedIndexes = IndexSet()
```
Инициализацию коллекций insertedIndexes, deletedIndexes, updatedIndexes можно вынести в один метод
Создать приватный метод:
```
private func resetChangeTracking() {
    insertedIndexes = IndexSet()
    deletedIndexes = IndexSet()
    updatedIndexes = IndexSet()
}
```
И использовать его:
```
func controllerWillChangeContent(_ controller: NSFetchedResultsController<NSFetchRequestResult>) {
    resetChangeTracking()
}

IgorChumikov commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,175 @@
+                insertedIndexes?.insert(newIndexPath.item)
+            }
+        @unknown default:
+            break
```
Добавить assertionFailure, чтобы во время разработки видеть новые кейсы:
```
@unknown default:
    assertionFailure("NSFetchedResultsController: unknown case detected")
```
Если Apple добавит новый тип изменений — ты сразу узнаешь на этапе тестирования.

IgorChumikov commented on CoreData/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,175 @@
+    }
+    
+    // MARK: - Private Methods
+    private func convertCategoryFromCoreData(from categoryCoreData: TrackerCategoryCoreData) throws -> TrackerCategory {
```
Названия методов чуть почистить
```
private func convert(categoryCoreData: TrackerCategoryCoreData)
```
или
```
private func makeCategory(from coreData: TrackerCategoryCoreData)
```
Это стандартный Swift-стиль.

IgorChumikov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,132 @@
+    // MARK: - Public Properties
+    weak var delegate: TrackerRecordStoreDelegate?
+    
+    var completedTrackers: [TrackerRecord] {
```
Упрощение свойства completedTrackers

Проблема:
- try в map, обёрнутый в do-catch, как и в предыдущем коде, обнуляет весь массив при первой ошибке.
- Это потенциальная потеря данных даже если большинство объектов валидные.

Как лучше:
Использовать .compactMap { try? }, чтобы пропускать только битые объекты:
```
var completedTrackers: [TrackerRecord] {
    (fetchedResultsController.fetchedObjects ?? []).compactMap {
        try? convertTrackerRecordFromCoreData(from: $0)
    }
}

IgorChumikov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,132 @@
+        
+        fetchedResultsController.delegate = self
+        
+        try? fetchedResultsController.performFetch()
```
Проблема: Ошибки игнорируются.

Лучше хотя бы залогировать ошибку в debug-сборках.

Как лучше:
```
do {
    try fetchedResultsController.performFetch()
} catch {
    assertionFailure("Failed to fetch tracker records: \(error)")
}
```
Или оставить try?, но добавить комментарий, чтобы явно показать намерение.

IgorChumikov commented on CoreData/TrackerRecordStore.swift:
```diff
@@ -0,0 +1,132 @@
+    }
+    
+    // MARK: - Public Methods
+    func addRecord(trackerId: UUID, date: Date) throws {
```
Оптимизация методов addRecord и deleteRecord
В addRecord
Сейчас ты создаёшь внутри каждый раз:
```
let trackerStore = TrackerStore(context: context)
```
Проблема:
- Ты создаёшь новый TrackerStore каждый раз при добавлении записи — это неэффективно.

Решение:
- Лучше передавать TrackerCoreData напрямую извне, либо держать TrackerStore внутри TrackerRecordStore как приватное свойство.

Или хотя бы принимать tracker в addRecord, чтобы не искать его заново.

IgorChumikov commented on Tracker_app/Weekdays.swift:
```diff
@@ -0,0 +1,51 @@
+
+enum WeekDays: Int, CaseIterable {
+    case monday = 1
+    case tueseday
```
Опечатка в названии case'а 
 - case tuesday

IgorChumikov commented on Work applications/CreationTrackerViewController.swift:
```diff
@@ -0,0 +1,276 @@
+}
+// MARK: - Types
+
+class CreationTrackerViewController: UIViewController {
```
final

IgorChumikov commented on Work applications/CreationTrackerViewController.swift:
```diff
@@ -0,0 +1,276 @@
+import UIKit
```
Этого файла нет в проекте

roudYStorm committed:
12345

roudYStorm committed:
р

roudYStorm committed:
переписанный проект

roudYStorm commented:
переделала весь проект с наставником
сама не разобралась
не удивляйтесь пожалуйста что много что я переписала можно сказать все

roudYStorm committed:
переписала проект, вместе с наставником заново чтобы разобраться
он не похож на 14 спринт но я нанимала наставника и мы вместе все переписали поэтому код отличается
